### PR TITLE
Fix Issue 20353 - -checkaction=context does not work well with const …

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
…numbers

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -5854,7 +5851,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                     auto binExp = cast(EqualExp) exp.e1;
 
                     // template args
-                    Expression comp = new StringExp(loc, Token.toString(exp.e1.op));
+                    Expression comp = new StringExp(loc, Token.toString(tok));
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d]: This change is unrelated and it also may be unsafe (I haven't verified, but it's possible that exp.e1 is rewritten to something else thus changing the token). I suggest doing this in a separate PR to make rusre everything goes well; also it is good practice to separate refactoring PRs from feature development PRs

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -5879,6 +5876,15 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
             }
         }
+
+        // https://issues.dlang.org/show_bug.cgi?id=20353
+        // Do not optimize before generating the message as the condition
+        // may be const-folded into a boolean.
+        //
+        // BUG: see if we can do compile time elimination of the Assert
+        exp.e1 = exp.e1.optimize(WANTvalue);
+        exp.e1 = exp.e1.toBoolean(sc);
```
Comment (RazvanN7) [inline src/dmd/expressionsem.d]: This change should be done only if the checkaction switch is enabled. If it's not, the current behavior should be kept. This may also get rid of the tester failures.

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/expressionsem.d]: I'll revert it. The patch was working locally and this was the only discarded change

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (MoonlightSentinel) [reply src/dmd/expressionsem.d]: This code above is only executed for `checkaction=context` so the code path is not changed without it.
```
/* Optimize */
if (false) { /* Rewrite message */ }
```
to 

```
if (false) { /* Rewrite message */ }
/* Optimize */
```

**Comment by MoonlightSentinel** [comment]
The test fails for release builds and vanishes when the `optimize` call is disabled (see last commit).
The failure is caused by an `ud2` in the function prolog of `testFolded()`: 
```
000000000000124c <_D10testassert10testFoldedFZv>:
    124c:	55                   	push   %rbp
    124d:	48 8b ec             	mov    %rsp,%rbp
    1250:	48 83 ec 10          	sub    $0x10,%rsp
    1254:	0f 0b                	ud2    
    1256:	eb 32                	jmp    128a <_D10testassert10testFoldedFZv+0x3e>
    1258:	48 8d a5 f0 ff ff ff 	lea    -0x10(%rbp),%rsp
    125f:	89 55 f8             	mov    %edx,-0x8(%rbp)
    1262:	48 89 c7             	mov    %rax,%rdi
    1265:	e8 46 fe ff ff       	callq  10b0 <__dmd_begin_catch@plt>
    126a:	48 89 45 f0          	mov    %rax,-0x10(%rbp)
    126e:	8b 45 f8             	mov    -0x8(%rbp),%eax
    1271:	83 f8 01             	cmp    $0x1,%eax
    1274:	74 02                	je     1278 <_D10testassert10testFoldedFZv+0x2c>
    1276:	eb 0e                	jmp    1286 <_D10testassert10testFoldedFZv+0x3a>
    1278:	48 8b 4d f0          	mov    -0x10(%rbp),%rcx
    127c:	48 8b 51 18          	mov    0x18(%rcx),%rdx
    1280:	48 8b 41 10          	mov    0x10(%rcx),%rax
    1284:	eb 04                	jmp    128a <_D10testassert10testFoldedFZv+0x3e>
    1286:	0f 0b                	ud2    
    1288:	c9                   	leaveq 
    1289:	c3                   	retq   
    128a:	c9                   	leaveq 
    128b:	c3                   	retq  
```
Is this a backend bug or caused by an malformed AST (which should probably abort compilation)?

**Comment by MoonlightSentinel** [comment]
The delayed `optimize` const-folds `x < y` into `false` in release mode, resulting in an `assert(false)` which aborts the test.
I'm not sure wether this is comformant to the [spec](https://dlang.org/spec/expression.html#assert_expressions):

> If the first AssignExpression consists entirely of compile time constants, and evaluates to false, it is a special case; it signifies that it is unreachable code. Compile Time Function Execution (CTFE) is not attempted.

CC @RazvanN7

**Comment by MoonlightSentinel** [comment]
This should be good now

**Comment by Geod24** [comment]
I'm not really convinced by this diff TBH. From what I see, this changes behavior so that constfolded expressions do not result in `assert(false)` anymore, which is kept even in `-release` builds.

And the benefit of this is just a slightly better output. Now it seems that every asserts that would be affected by this bug could be turned into a `static assert`, for which we have been generating user-friendly message for a while now:
```D
void main() {
    const x = -1L;
    const y = -2L;
    static assert(x == y);
}
```

```
pr10535.d(4): Error: static assert:  -1L == -2L is false
```

So, why not re-use this mechanism instead and provide the string directly from the compiler ?

**Comment by MoonlightSentinel** [comment]
> I'm not really convinced by this diff TBH. From what I see, this changes behavior so that constfolded expressions do not result in `assert(false)` anymore, which is kept even in `-release` builds.

The example in the test is constfolded and treated as `assert(false)` (hence its excluded if `testassert.d` is compiled with `-release`). Do you have a concrete example where the behavoir differs from `master`?

> And the benefit of this is just a slightly better output. Now it seems that every asserts that would be affected by this bug could be turned into a `static assert`, for which we have been generating user-friendly message for a while now:
> 
> ```d
> void main() {
>     const x = -1L;
>     const y = -2L;
>     static assert(x == y);
> }
> ```
> 
> ```
> pr10535.d(4): Error: static assert:  -1L == -2L is false
> ```
> 
> So, why not re-use this mechanism instead and provide the string directly from the compiler ?

Are you suggesting to implicitly promote `assert(cond)` to `static assert(cond)` if `cond`can be evaluated at compile time?

**Comment by Geod24** [comment]
> hence its excluded if `testassert.d` is compiled with `-release`

`assert(false)` are never removed, even in `-release` mode.
They are used to mark unreachable code.

```console
% echo "void main () { const x = 1; const y = 42; assert(x == y); }" > pr10535.d
% dmd -release -run pr10535.d
Error: program killed by signal 4
```

That being said, I just tested this PR and the `assert` is still there for `-release` builds, with or without `-checkaction=context`, so that's good.

> Are you suggesting to implicitly promote `assert(cond)` to `static assert(cond)` if cond is non-trivial (`0`, `false`) but can be evaluated at compile time?

No, as that would break the usage of "`assert(0)` for unreachable code". What I was suggesting is to just generate the message upfront, in DMD, when the `assert` is const-folded, instead of relying on druntime.

**Comment by MoonlightSentinel** [comment]
> > hence its excluded if `testassert.d` is compiled with `-release`
> 
> `assert(false)` are never removed, even in `-release` mode.
> They are used to mark unreachable code.
> 
> ```
> % echo "void main () { const x = 1; const y = 42; assert(x == y); }" > pr10535.d
> % dmd -release -run pr10535.d
> Error: program killed by signal 4
> ```
> 
> That being said, I just tested this PR and the `assert` is still there for `-release` builds, with or without `-checkaction=context`, so that's good.

I think my previous reply was worded poorly- the test case (not the `assert`) is excluded when compiling with `-release`.


> No, as that would break the usage of "`assert(0)` for unreachable code". What I was suggesting is to just generate the message upfront, in DMD, when the `assert` is const-folded, instead of relying on druntime.

Thats could be an improvement but would require special-casing depending on the operands as `__d_assert_fail` can format more expression than `toChars` used by `static assert`. (unless we can assume that these expressions cannot be constfolded by the fronted)

I'm not entirely sure why your example even works but `dmd` seems to promote `x`, `y` to `enums`.

**Comment by Geod24** [comment]
> Thats could be an improvement but would require special-casing depending on [...]

I was thinking of just providing a template overload that accepts a string. Do you h

Can you add the following test case to your PR:
```D
struct Foo { int a; float b; }
static immutable Foo x, y;
void main ()
{
    assert(x == y);
}
```

For me this still print `core.exception.AssertError@pr10535.d(5): assert(x == y) failed` with your branch.
Interestingly, when the struct only have a single `float` member, it properly prints "nan != nan" (even without your PR).

**Comment by MoonlightSentinel** [comment]
> > Thats could be an improvement but would require special-casing depending on [...]
> 
> I was thinking of just providing a template overload that accepts a string. Do you h

There is something missing here?


> Can you add the following test case to your PR:
> 
> ```d
> struct Foo { int a; float b; }
> static immutable Foo x, y;
> void main ()
> {
>     assert(x == y);
> }
> ```

Done


> For me this still print `core.exception.AssertError@pr10535.d(5): assert(x == y) failed` with your branch.

`unaSemantic` rewrites that assert to `assert(x.a == y.a && (x.b == y.b))`


> Interestingly, when the struct only have a single `float` member, it properly prints "nan != nan" (even without your PR).

`unaSemantic` rewrites that assert to `assert(x.b == y.b)`


Shouldn't these assert generate an implicit `opEquals` and be rewritten into `assert(x.opEquals(y))`?

**Comment by MoonlightSentinel** [comment]
> > For me this still print `core.exception.AssertError@pr10535.d(5): assert(x == y) failed` with your branch.
> 
> `unaSemantic` rewrites that assert to `assert(x.a == y.a && (x.b == y.b))`

Essentially this is a problem of the current implementation of `checkaction=context` not my PR as it is unable to handle complex expressions.

The correct approach IMHO would be to apply `checkaction=context` before rewriting to format the original expression instead of the lowering. This would require less special casing and produce better messages (`S(0, 1) != S(1, 1)` instead of maybe `0 != 1 || 1 != 1`).

Handling complex expressions would also suffice to handle such `==` lowerings at the cost of slightly worse messages.

I tried to implement the first approach but was not successfull yet

**Review by RazvanN7 — APPROVED** [review]
I think that the fix is fine. I have hit this problem in the past also where the optimizer kicks in way too early, before actual error checking is done, and by the time you want to print the error the AST has already been irreversibly altered.

**Comment by MoonlightSentinel** [comment]
This issue will probably be fixed by #11005
